File: backend/app/schemas/instagram.py
import datetime
import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.meme_container import ContainerMetadataStatus, ContainerPlatform
from app.schemas.auth import UserOut
from app.schemas.memes import MemeOut


class MemeContainerCreate(BaseModel):
    source_url: str = Field(min_length=1, max_length=2048)


class MemeContainerOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    submitter: UserOut
    platform: ContainerPlatform
    source_url: str
    title: str | None
    thumbnail_url: str | None
    metadata_status: ContainerMetadataStatus
    upvote_count: int
    downvote_count: int
    score: int
    comment_count: int
    # Private engagement data — null unless the caller is the submitter. See
    # services/instagram.py::_build_container_out.
    view_count: int | None
    viewer_vote: Literal[1, -1] | None
    created_at: datetime.datetime


class ContainerViewOut(BaseModel):
    meme_container_id: uuid.UUID
    view_count: int


class ContainerCommentCreate(BaseModel):
    body: str = Field(min_length=1, max_length=500)


class ContainerCommentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    author: UserOut
    body: str
    created_at: datetime.datetime


class MemeFeedItem(BaseModel):
    kind: Literal["meme"]
    meme: MemeOut


class ContainerFeedItem(BaseModel):
    kind: Literal["container"]
    container: MemeContainerOut


class MergedFeedPage(BaseModel):
    """The public feed's actual response shape: native `Meme`s and `MemeContainer`s
    (externally-shared Reels/posts) merged and sorted together — the two content types
    share one feed by design (confirmed with user), even though they're backed by
    entirely separate tables/reactions/comments/votes under the hood. Ranked by Reddit-
    style Hot score (vote score vs. age), not recency — offset-paginated (`has_more`),
    since Hot score drifts continuously with time and has no stable keyset cursor.
    """

    items: list[MemeFeedItem | ContainerFeedItem]
    has_more: bool

File: frontend/src/services/instagram.ts
import { api } from '@/services/api';
import type { AuthUserResponse } from '@/services/auth';

export type ContainerMetadataStatus = 'pending' | 'ready' | 'failed';

export interface MemeContainerResponse {
  id: string;
  submitter: AuthUserResponse;
  platform: 'instagram';
  source_url: string;
  title: string | null;
  thumbnail_url: string | null;
  metadata_status: ContainerMetadataStatus;
  upvote_count: number;
  downvote_count: number;
  score: number;
  comment_count: number;
  // Private engagement data — null unless the viewer is the submitter. See backend
  // services/instagram.py::_build_container_out.
  view_count: number | null;
  viewer_vote: 1 | -1 | null;
  created_at: string;
}

export interface ContainerViewResponse {
  meme_container_id: string;
  view_count: number;
}

export interface ContainerCommentResponse {
  id: string;
  author: AuthUserResponse;
  body: string;
  created_at: string;
}

export function createContainerRequest(sourceUrl: string) {
  return api.post<MemeContainerResponse>('/instagram/containers', { source_url: sourceUrl });
}

export function getContainerRequest(containerId: string) {
  return api.get<MemeContainerResponse>(`/instagram/containers/${containerId}`);
}

export function castContainerVoteRequest(containerId: string, value: 1 | -1) {
  return api.post<MemeContainerResponse>(`/instagram/containers/${containerId}/votes`, { value });
}

// Records one impression on a container — reach signal for its MemeScore. Fire-and-forget.
export function recordContainerViewRequest(containerId: string) {
  return api.post<ContainerViewResponse>(`/instagram/containers/${containerId}/views`);
}

export function listContainerCommentsRequest(containerId: string) {
  return api.get<ContainerCommentResponse[]>(`/instagram/containers/${containerId}/comments`);
}

export function addContainerCommentRequest(containerId: string, body: string) {
  return api.post<ContainerCommentResponse>(`/instagram/containers/${containerId}/comments`, {
    body,
  });
}
